### ISSUE TITLE
AttributeError: 'mpz' object has no attribute 'denominator' with sp.series()

### ISSUE THREAD

**Issue by magv** [sympy/sympy#21859]
Hi, all. I've run into a problem that can be traced back to this example:

```python
import sympy as sp
expr = sp.sympify("-exp(-n*log(z)/2)*exp(-5*I*pi*n/6)*gamma(5*n/6 + 1)/(gamma(n/3 + 1)*gamma(n/2 + 1))")    
var = sp.sympify("n")
generator = sp.series(expr, var, n=None)                                                                    
print(next(generator))                                                                                      
print(next(generator))                                                                                      
```
If you'll run this, it will fail with this output:
```
-1
Traceback (most recent call last):
  File "example.py", line 7, in <module>
    print(next(generator))
  File "sympy/core/expr.py", line 2985, in <genexpr>
    return (s.subs(xpos, x) for s in rv)
  File "sympy/core/expr.py", line 3048, in yield_lseries
    o = Order(si, x)*x
  File "sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/series/order.py", line 201, in __new__
    expr = expr.factor()
  File "sympy/core/expr.py", line 3702, in factor
    return factor(self, *gens, **args)
  File "sympy/polys/polytools.py", line 6390, in factor
    return _generic_factor(f, gens, args, method='factor')
  File "sympy/polys/polytools.py", line 6062, in _generic_factor
    return _symbolic_factor(sympify(expr), opt, method)
  File "sympy/polys/polytools.py", line 6002, in _symbolic_factor
    coeff, factors = _symbolic_factor_list(together(expr, fraction=opt['fraction']), opt, method)
  File "sympy/polys/polytools.py", line 5967, in _symbolic_factor_list
    _coeff, _factors = func()
  File "sympy/polys/polytools.py", line 3336, in factor_list
    coeff, factors = f.rep.factor_list()
  File "sympy/polys/polyclasses.py", line 795, in factor_list
    coeff, factors = dmp_factor_list(f.rep, f.lev, f.dom)
  File "sympy/polys/factortools.py", line 1391, in dmp_factor_list
    coeff, factors = dmp_zz_i_factor(f, u, K0)
  File "sympy/polys/factortools.py", line 1213, in dmp_zz_i_factor
    content, fac_prim = dmp_ground_primitive(fac_num_ZZ_I, u, K1)
  File "sympy/polys/densetools.py", line 719, in dmp_ground_primitive
    cont = dmp_ground_content(f, u, K)
  File "sympy/polys/densetools.py", line 650, in dmp_ground_content
    cont = K.gcd(cont, dmp_ground_content(c, v, K))
  File "sympy/polys/densetools.py", line 638, in dmp_ground_content
    return dup_content(f, K)
  File "sympy/polys/densetools.py", line 605, in dup_content
    cont = K.gcd(cont, c)
  File "sympy/polys/domains/field.py", line 66, in gcd
    p = ring.gcd(self.numer(a), self.numer(b))
  File "sympy/polys/domains/gaussiandomains.py", line 612, in numer
    return ZZ_I.convert(a * self.denom(a))
  File "sympy/polys/domains/gaussiandomains.py", line 619, in denom
    denom_ZZ = ZZ.lcm(QQ.denom(a.x), QQ.denom(a.y))
  File "sympy/polys/domains/gmpyrationalfield.py", line 96, in denom
    return a.denominator
AttributeError: 'mpz' object has no attribute 'denominator'
```

Now the cause of this problem seems to be the difference between gmpy and gmpy2 libraries: it seems that I have gmpy 1.17 installed (from system packages), but no gmpy2, and this code seems to assume interface that only exists in gmpy2: there is no `denominator` property in a `ZZ(1)` with gmpy, only with gmpy2.

I can actually confirm that once I install gmpy2 (from PyPI) everything works. So the question is: why doesn't SymPy simply require gmpy2? I mean, otherwise you'll need to make sure both APIs are supported, which is some non-trivial work, so I'm guessing there's a good reason for this.

Anyway, an additional problem I have is that this failure actually affects [pySecDec](https://github.com/gudrunhe/secdec), so I need some sort of a solution not just for me but for our users too. I'm tempted to just add gmpy2 to the list of pySecDec dependencies, to be honest, but I'd like to hear your thoughts about it first.

**Comment by oscarbenjamin** [comment]
Support for using gmpy (rather than gmpy2) is going to be dropped in SymPy 1.9 (#21016).

**Comment by oscarbenjamin** [comment]
@magv does that answer your question?

I think this should probably be closed as wontfix in sympy since support for gmpy is to be dropped in the next release.

**Comment by magv** [comment]
My takeaway is that for the moment I should just add gmpy2 to the list of dependencies along with sympy. If this takeaway is correct, then yes, close as "resolved in 1.8.1".

**Comment by oscarbenjamin** [comment]
There won't be a 1.8.1 but there will be a 1.9 release soon and in that release SymPy will no longer attempt to use gmpy if it is installed. It isn't necessary to have gmpy or gmpy2 with any release of SymPy (these are optional dependencies for speed).

If pySecDec wants to support running with older SymPy releases then it could add gmpy2 as a dependency which would mean that even if the user has installed gmpy then gmpy2 would be used instead. Alternatively soon there will be a SymPy 1.9 and pySecDec could require `sympy >= 1.9`.

I'm surprised this hasn't been noticed before since the issue has been around since sympy 1.7 (released in November) and can be triggered easily:
```python
In [4]: factor(x + I)
---------------------------------------------------------------------------
AttributeError: 'mpz' object has no attribute 'denominator'
```
I think the fact that this hasn't come up shows that few people are using gmpy (rather than gmpy2). Also gmpy was not being tested in SymPy's own CI.
